File: 4_Lookback Option/Option_Lookback_CRR.py
from math import log, sqrt, exp
import bisect
import time

def simulate_calibrated_stock_price(St, u, d, layers):
    stockPrice = []
    for i in range(layers+1):
        stockPrice.append([0]*(i+1))
    for i in range(layers+1):
        for j in range(i+1):
            stockPrice[i][j] = St * u**(i-j) * d**j

    last1 = stockPrice[-1]
    last2 = stockPrice[-2]
    calibration = sorted(last1 + last2, reverse = True)
    calibration[layers] = St

    caliIndex = []
    for i in range(layers,-layers-1,-1):
        caliIndex.append(i)
    calibration = dict(zip(caliIndex, calibration))

    # indexDiff : u的次方 - d的次方
    for i in range(layers+1):
        for j in range(i+1):
            indexDiff = (i-j) - j
            stockPrice[i][j] = calibration[indexDiff]

    return stockPrice


class Tree_Node:
    def __init__(self, St):
        self.St = St
        self.SmaxLst = []
        self.payoff = []
        self.putValue = []


def lookback_CRR(StMax, St, T, r, q, sigma, layers, type):
    dt = T/layers
    u = exp(sigma * sqrt(dt))
    d = exp(-sigma * sqrt(dt))
    p = (exp((r-q)*dt) - d)/(u - d)

    stockPrice = simulate_calibrated_stock_price(St, u, d, layers)

    # 建立節點
    TreeNodes = []
    for i in range(layers+1):
        TreeNodes.append([0]* (i+1))
    for i in range(layers+1):
        for j in range(i+1):
            TreeNodes[i][j] = Tree_Node(stockPrice[i][j])

    # 第0個節點 (即今日)
    if StMax >= St:
        TreeNodes[0][0].SmaxLst = [StMax]
    else:
        TreeNodes[0][0].SmaxLst = [St]

    # Build Smax by forward tracking method
    for i in range(1, layers+1):
        for j in range(i+1):
            if j == 0:
                for Smax in TreeNodes[i-1][j].SmaxLst:
                    if Smax >= TreeNodes[i][j].St:
                        if Smax not in TreeNodes[i][j].SmaxLst:
                            bisect.insort(TreeNodes[i][j].SmaxLst, Smax)
                    else:
                        if TreeNodes[i][j].St not in TreeNodes[i][j].SmaxLst:
                            bisect.insort(TreeNodes[i][j].SmaxLst, TreeNodes[i][j].St)
            
            elif j == i:
                for Smax in TreeNodes[i-1][j-1].SmaxLst:
                    if Smax >= TreeNodes[i][j].St:
                        if Smax not in TreeNodes[i][j].SmaxLst:
                            bisect.insort(TreeNodes[i][j].SmaxLst, Smax)
                    else:
                        if TreeNodes[i][j].St not in TreeNodes[i][j].SmaxLst:
                            bisect.insort(TreeNodes[i][j].SmaxLst, TreeNodes[i][j].St)
            
            else:
                for Smax in TreeNodes[i-1][j-1].SmaxLst:
                    if Smax >= TreeNodes[i][j].St:
                        if Smax not in TreeNodes[i][j].SmaxLst:
                            bisect.insort(TreeNodes[i][j].SmaxLst, Smax)
                    else:
                        if TreeNodes[i][j].St not in TreeNodes[i][j].SmaxLst:
                            bisect.insort(TreeNodes[i][j].SmaxLst, TreeNodes[i][j].St)

                for Smax in TreeNodes[i-1][j].SmaxLst:
                    if Smax >= TreeNodes[i][j].St:
                        if Smax not in TreeNodes[i][j].SmaxLst:
                            bisect.insort(TreeNodes[i][j].SmaxLst, Smax)
                    else:
                        if TreeNodes[i][j].St not in TreeNodes[i][j].SmaxLst:
                            bisect.insort(TreeNodes[i][j].SmaxLst, TreeNodes[i][j].St)
            
            # 最後一層要有payoff list
            if i == layers:
                # 現在，我有Smax list
                for Smax in TreeNodes[i][j].SmaxLst:
                        TreeNodes[i][j].putValue.append(max(Smax-TreeNodes[i][j].St, 0))

    # Backward induction
    times = 0
    i_temp = layers - 1
    while times < layers:
        for j in range(i_temp+1):
            ku, kd = 0, 0
            for Smax in TreeNodes[i_temp][j].SmaxLst:
                u_is_found = False
                # search for ku in the SmaxLst of the upper node in next layer
                # search for self
                for k in range(ku, len(TreeNodes[i_temp+1][j].SmaxLst)):
                    if TreeNodes[i_temp+1][j].SmaxLst[k] == Smax:
                        ku = k
                        u_is_found = True
                        break
                # search for self*u
                if u_is_found == False:
                    for k in range(ku, len(TreeNodes[i_temp+1][j].SmaxLst)):
                        if abs(TreeNodes[i_temp+1][j].SmaxLst[k] - Smax*u) < 10**-8:
                            ku = k
                            break

                # search for kd in the SmaxLst of the lower node in next layer
                for k in range(kd, len(TreeNodes[i_temp+1][j+1].SmaxLst)):
                    if TreeNodes[i_temp+1][j+1].SmaxLst[k] == Smax:
                        kd = k
                        break

                # using index() method
                # if Smax in TreeNodes[i_temp+1][j].SmaxLst:
                #     ku = TreeNodes[i_temp+1][j].SmaxLst.index(Smax)
                # else:
                #     for k in range(len(TreeNodes[i_temp+1][j].SmaxLst)):
                #         if abs(TreeNodes[i_temp+1][j].SmaxLst[k] - Smax) < 10**-8:
                #             ku = TreeNodes[i_temp+1][j].SmaxLst.index(Smax)
                #             break
                
                # kd = TreeNodes[i_temp+1][j+1].SmaxLst.index(Smax)

                discounted = (TreeNodes[i_temp+1][j].putValue[ku]*p + TreeNodes[i_temp+1][j+1].putValue[kd]*(1-p)) * exp(-r*dt)
                if type == 'American':
                    discounted = max(Smax - TreeNodes[i_temp][j].St, discounted)
                TreeNodes[i_temp][j].putValue.append(discounted)

        i_temp -= 1
        times += 1

    putValue = round(TreeNodes[0][0].putValue[0], 4)
    print(f'(CRR Binomial Tree) Price of {type} Lookback Put : {putValue}')
    return putValue



# main
St = 50
T = 0.25
r = 0.1
q = 0
sigma = 0.4

start = time.perf_counter()

print('============================================================')
StMax = 50
print('Lookback Option')
print(f'[ Smax,t = {StMax} ]')
print('------------------------------------------------------------')
layers = 100
print(f'n = {layers}')
lookback_CRR(StMax, St, T, r, q, sigma, layers, 'European')
lookback_CRR(StMax, St, T, r, q, sigma, layers, 'American')
print('------------------------------------------------------------')
layers = 300
print(f'n = {layers}')
lookback_CRR(StMax, St, T, r, q, sigma, layers, 'European')
lookback_CRR(StMax, St, T, r, q, sigma, layers, 'American')
print()

print('============================================================')
StMax = 60
print('Lookback Option')
print(f'[ Smax,t = {StMax} ]')
print('------------------------------------------------------------')
layers = 100
print(f'n = {layers}')
lookback_CRR(StMax, St, T, r, q, sigma, layers, 'European')
lookback_CRR(StMax, St, T, r, q, sigma, layers, 'American')
print('------------------------------------------------------------')
layers = 300
print(f'n = {layers}')
lookback_CRR(StMax, St, T, r, q, sigma, layers, 'European')
lookback_CRR(StMax, St, T, r, q, sigma, layers, 'American')
print()

print('============================================================')
StMax = 70
print('Lookback Option')
print(f'[ Smax,t = {StMax} ]')
print('------------------------------------------------------------')
layers = 100
print(f'n = {layers}')
lookback_CRR(StMax, St, T, r, q, sigma, layers, 'European')
lookback_CRR(StMax, St, T, r, q, sigma, layers, 'American')
print('------------------------------------------------------------')
layers = 300
print(f'n = {layers}')
lookback_CRR(StMax, St, T, r, q, sigma, layers, 'European')
lookback_CRR(StMax, St, T, r, q, sigma, layers, 'American')
print()

finish = time.perf_counter()
print(f'Process finished in {round(finish - start, 2)} second(s).')
print('\n')

File: 4_Lookback Option/Option_Lookback_CRR.cpp
# include <iostream>
# include <vector>
# include <map>
# include <cmath>
# include <algorithm>

using namespace std;

vector<vector<double>> simulate_calibrated_stock_price(double St, double u, double d, int layers){
    vector<vector<double>> stockPrice;
    for(int i = 0; i < layers+1; i++){
        vector<double> temp;
        for(int j = 0; j < i+1; j++){
            temp.push_back(0);
        }
        stockPrice.push_back(temp);
    }
    for(int i = 0; i < layers+1; i++){
        for(int j = 0; j < i+1; j++){
            stockPrice[i][j] = St * pow(u, i-j) * pow(d, j);
        }
    }
    vector<double> last1 = stockPrice.end()[-1];
    vector<double> last2 = stockPrice.end()[-2];

    // use last1 for price calibration
    // insert last 2 into last 1 and sort
    last1.insert(last1.end(), last2.begin(), last2.end());
    sort(last1.begin(), last1.end(), greater<double>());
    last1[layers] = St;

    // build the vector for calibration
    vector<int> caliIndex;
    for(int i = layers; i > -layers-1; i -= 1){
        caliIndex.push_back(i);
    }
    map<int, double> calibration;
    for(int i = 0; i < layers*2+1; i++){
        calibration[caliIndex[i]] = last1[i];
    }
    for(int i = layers; i > -layers-1; i -= 1){
        cout << calibration[i] << ",";
    }

    // indexDiff : power of u - power of d
    for(int i = 0; i < layers+1; i++){
        for(int j = 0; j < i+1; j++){
            int indexDiff = (i-j) - j;
            stockPrice[i][j] = calibration[indexDiff];
        }
    }

    return stockPrice;
}

class Tree_Node{
    public:
    double St;
    vector<double> SmaxLst;
    vector<double> optionValue;
    
    // constructor
    Tree_Node(double St_ij){
        St = St_ij;
    }
};

// double lookback_CRR_put(double StMax, double St, double T, double r, double q, double sigma, int layers, string type){
//     double dt = T/layers;
//     double u = exp(sigma*sqrt(dt));
//     double d = exp(-sigma*sqrt(dt));
//     double p = (exp((r-q)*dt) - d)/(u - d);

//     vector<vector<double>> stockPrice = simulate_calibrated_stock_price(St, u, d, layers);

//     // build Nodes
//     return;
// }


int main(){
    double St = 50;
    double K = 50;
    double r = 0.1;
    double q = 0.05;
    double sigma = 0.4;
    double T = 0.5;
    int layers = 4;

    double dt = T/layers;
    double u = exp(sigma*sqrt(dt));
    double d = exp(-sigma*sqrt(dt));
    double p = (exp((r-q)*dt) - d)/(u - d);

    vector<vector<double>> stockPrice = simulate_calibrated_stock_price(St, u, d, layers);
    for(int i = 0; i < layers+1; i++){
        for(int j = 0; j < i+1; j++){
            cout << stockPrice[i][j] << "," ;
        }
        cout << endl;
    }

    return 0;
}